File: Arrays/Array_Rotation/Array_Rotation.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Array_Rotation
{

    int gcd(int a,int b)
    {
        if(b==0) {
            return a;
        }
        return gcd(b, a%b);
    }

    void rotateLeft(Integer[] arr, int n, int r)
    {
        int gcd = gcd(n, r);
        for(int i=0 ; i<gcd ; i++)
        {
            int temp = arr[i];
            int j = i;
            while(true)
            {
                int k = j + r;
                if(k >= n) {
                    k -= n;
                }
                if(k == i) {
                    break;
                }   
                arr[j] = arr[k];
                j = k;
            }
            arr[j] = temp;
        }
    }

    void rotateRight(Integer[] arr, int n, int r)
    {
        int gcd = gcd(n,r);
        for(int i=0; i<gcd ; i++)
        {
            int temp = arr[i];
            int j = i;
            while(true)
            {
                int k = j - r;
                if(k < 0) {
                    k += n;
                }
                if(k == i) {
                    break;
                }
                arr[j] = arr[k];
                j = k;
            }
            arr[j] = temp;
        }
    }

    public static void main(String[] args)
    {
        Array_Rotation arrayRotation = new Array_Rotation();
        Integer[] array = {1, 2, 3, 4, 5, 6, 7};
        
        arrayRotation.rotateLeft(array, 7, 3);
        System.out.println(Arrays.toString(array));
        
        arrayRotation.rotateRight(array, 7, 4);
        System.out.println(Arrays.toString(array));
    }
};

File: Arrays/Array_Rotation/Array_Rotation.cpp
#include <cstdio>
using namespace std;

// TO-DO : Fix this code

int gcd(int a, int b)
{
    if (b == 0)
        return a;
    else
        return gcd(b, a % b);
}

void rotateArray(int arr[], int n, int d)
{
    int g = gcd(n, d);
    int temp;
    int j;
    if (g == 1)
    {
        
        for (int i = 0; i < g; i++)
        {
            temp = arr[i];
            j = i;
            while (1)
            {
                if (j + d < n)
                {
                    arr[j] = arr[j + d];
                    j += d;
                }
                else
                {
                    arr[n-d+i] = temp;
                    break;
                }
            }
        }
    }
    else
    {
        for (int i = 0; i < g; i++)
        {
            temp = arr[i];
            j = i;
            while (1)
            {
                if (j + d < n)
                {
                    arr[j] = arr[j + d];
                    j += d;
                }
                else
                {
                    arr[j] = temp;
                    break;
                }
            }
        }
    }
}

void printArray(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main()
{
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13};
    int d = 3;
    int n = sizeof(arr) / sizeof(int);
    printArray(arr, n);
    rotateArray(arr, n, d);
    printArray(arr, n);
    return 0;
}